File: CS/EditableGroupRow/GroupRowEdit.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Drawing;
using System.Windows.Forms;

using DevExpress.Utils.Menu;
using DevExpress.XtraGrid;
using DevExpress.XtraGrid.Columns;
using DevExpress.XtraEditors;
using DevExpress.XtraGrid.Views;
using DevExpress.XtraGrid.Views.Grid;
using DevExpress.XtraGrid.Views.Grid.ViewInfo;

namespace EditableGroupRow
{
    public class GroupRowEdit
    {
        int rowToChangeHandle = GridControl.InvalidRowHandle;
        GridView View;
        public GroupRowEdit(DevExpress.XtraGrid.Views.Grid.GridView view)
        { 
            View = view;
            View.PopupMenuShowing +=new PopupMenuShowingEventHandler(View_PopupMenuShowing);
        }

        private void View_PopupMenuShowing(object sender, DevExpress.XtraGrid.Views.Grid.PopupMenuShowingEventArgs e)
        {
            if (View.IsGroupRow(e.HitInfo.RowHandle))
            {
                e.Menu.Items.Clear();
                DXMenuItem item = new DXMenuItem("Edit");
                item.Click += new EventHandler(item_Click);
                e.Menu.Items.Add(item);
            }
        }
        void item_Click(object sender, EventArgs e)
        {
            ShowGroupEdit();
        }

        void ShowGroupEdit()
        {
            TextEdit te = new TextEdit();
            this.View.GridControl.Controls.Add( te);
            rowToChangeHandle = this.View.FocusedRowHandle;
            GridViewInfo gvi = new GridViewInfo(this.View);
            Graphics g = this.View.GridControl.CreateGraphics();
            gvi.Calc(g, this.View.ViewRect);
            GridGroupRowInfo info = gvi.RowsInfo.GetInfoByHandle(rowToChangeHandle) as GridGroupRowInfo;
            string newText = this.View.GetGroupRowValue(rowToChangeHandle).ToString();
            te.Text = newText;
            te.Bounds = info.DataBounds;
            te.Visible = true;
            te.Focus();
            te.SelectAll();
            te.Leave += new System.EventHandler(this.textEdit_Leave);
            te.KeyDown += new System.Windows.Forms.KeyEventHandler(this.textEdit_KeyDown);
        }
        void HideGroupEdit(object sender)
        {
                rowToChangeHandle = GridControl.InvalidRowHandle;
                (sender as TextEdit).Visible = false;
                View.GridControl.Controls.Remove(sender as TextEdit);
        }

        private void textEdit_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Enter)
            {
                UpdateNewRowString(sender);
            }
            if (e.KeyCode == Keys.Escape)
            {
                HideGroupEdit(sender);
            }
        }

        private void textEdit_Leave(object sender, EventArgs e)
        {
            UpdateNewRowString(sender);
        }

        void UpdateNewRowString(object sender)
        {
            string newValue = (sender as TextEdit).Text;
            View.BeginSort();
            GridColumn grColumn = View.GroupedColumns[View.GetRowLevel(rowToChangeHandle)];
            List<int> indexList = new List<int>();
            for (int i = 0; i < View.GetChildRowCount(rowToChangeHandle); i++)
            {
                int index = View.GetChildRowHandle(rowToChangeHandle, i);
                View.SetRowCellValue(index, grColumn, newValue);
            }
            View.EndSort();
            HideGroupEdit(sender);
        }
    }
}

File: VB/EditableGroupRow/GroupRowEdit.vb
Imports System
Imports System.Collections.Generic
Imports System.Drawing
Imports System.Windows.Forms
Imports DevExpress.Utils.Menu
Imports DevExpress.XtraGrid
Imports DevExpress.XtraGrid.Columns
Imports DevExpress.XtraEditors
Imports DevExpress.XtraGrid.Views.Grid
Imports DevExpress.XtraGrid.Views.Grid.ViewInfo

Namespace EditableGroupRow

    Public Class GroupRowEdit

        Private rowToChangeHandle As Integer = GridControl.InvalidRowHandle

        Private View As GridView

        Public Sub New(ByVal view As GridView)
            Me.View = view
            AddHandler Me.View.PopupMenuShowing, New PopupMenuShowingEventHandler(AddressOf View_PopupMenuShowing)
        End Sub

        Private Sub View_PopupMenuShowing(ByVal sender As Object, ByVal e As PopupMenuShowingEventArgs)
            If View.IsGroupRow(e.HitInfo.RowHandle) Then
                If e.Menu Is Nothing Then e.Menu = New DevExpress.XtraGrid.Menu.GridViewMenu(View)
                e.Menu.Items.Clear()
                Dim item As DXMenuItem = New DXMenuItem("Edit")
                AddHandler item.Click, New EventHandler(AddressOf item_Click)
                e.Menu.Items.Add(item)
            End If
        End Sub

        Private Sub item_Click(ByVal sender As Object, ByVal e As EventArgs)
            ShowGroupEdit()
        End Sub

        Private Sub ShowGroupEdit()
            Dim te As TextEdit = New TextEdit()
            View.GridControl.Controls.Add(te)
            rowToChangeHandle = View.FocusedRowHandle
            Dim gvi As GridViewInfo = New GridViewInfo(View)
            Dim g As Graphics = View.GridControl.CreateGraphics()
            gvi.Calc(g, View.ViewRect)
            Dim info As GridGroupRowInfo = TryCast(gvi.RowsInfo.GetInfoByHandle(rowToChangeHandle), GridGroupRowInfo)
            Dim newText As String = View.GetGroupRowValue(rowToChangeHandle).ToString()
            te.Text = newText
            te.Bounds = info.DataBounds
            te.Visible = True
            te.Focus()
            te.SelectAll()
            AddHandler te.Leave, New EventHandler(AddressOf textEdit_Leave)
            AddHandler te.KeyDown, New KeyEventHandler(AddressOf textEdit_KeyDown)
        End Sub

        Private Sub HideGroupEdit(ByVal sender As Object)
            rowToChangeHandle = GridControl.InvalidRowHandle
            TryCast(sender, TextEdit).Visible = False
            View.GridControl.Controls.Remove(TryCast(sender, TextEdit))
        End Sub

        Private Sub textEdit_KeyDown(ByVal sender As Object, ByVal e As KeyEventArgs)
            If e.KeyCode = Keys.Enter Then
                UpdateNewRowString(sender)
            End If

            If e.KeyCode = Keys.Escape Then
                HideGroupEdit(sender)
            End If
        End Sub

        Private Sub textEdit_Leave(ByVal sender As Object, ByVal e As EventArgs)
            UpdateNewRowString(sender)
        End Sub

        Private Sub UpdateNewRowString(ByVal sender As Object)
            Dim newValue As String = TryCast(sender, TextEdit).Text
            View.BeginSort()
            Dim grColumn As GridColumn = View.GroupedColumns(View.GetRowLevel(rowToChangeHandle))
            Dim indexList As List(Of Integer) = New List(Of Integer)()
            For i As Integer = 0 To View.GetChildRowCount(rowToChangeHandle) - 1
                Dim index As Integer = View.GetChildRowHandle(rowToChangeHandle, i)
                View.SetRowCellValue(index, grColumn, newValue)
            Next

            View.EndSort()
            HideGroupEdit(sender)
        End Sub
    End Class
End Namespace
